src/day9.scala

```diff
@@ -0,0 +1,112 @@
+// using scala 3.1.0
+
+package day9
+
+import scala.collection.immutable.Queue
+import scala.util.Using
+import scala.io.Source
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day9"))(_.mkString)
+
+type Height = Int
+case class Position(x: Int, y: Int)
+
+class Heightmap(val width: Int, val height: Int, private val data: Array[Height]):
+
+  def apply(coord: Position): Height = data(coord.y * width + coord.x)
+
+  def neighborsOf(coord: Position): Iterator[(Position, Height)] =
+    val Position(x, y) = coord
+    val idx = coord.y * width + coord.x
+    Iterator(
+      Option.when(x > 0)(Position(x - 1, y) -> data(idx - 1)),
+      Option.when(x < width - 1)(Position(x + 1, y) -> data(idx + 1)),
+      Option.when(y > 0)(Position(x, y - 1) -> data(idx - width)),
+      Option.when(y < height - 1)(Position(x, y + 1) -> data(idx + width))
+    ).flatMap(Iterator.from)
+ 
+  def lowPointsPositions: Iterator[Position] =
+    Iterator.range(0, data.length).map { idx =>
+      val x = idx % width
+      val y = idx / width
+      val coord = Position(x, y)
+      (
+        data(idx),
+        coord,
+        this.neighborsOf(coord).map(_._2)
+      )
+    }.collect {
+      case (value, coords, neighbors) if neighbors.forall(value < _) => 
+        coords
+    }
+end Heightmap
+
+ 
+object Heightmap:
+  def fromString(raw: String): Heightmap =
+    val width = raw.linesIterator.next.length
+    val height = raw.linesIterator.length
+    val data = raw.iterator.filter(_.isDigit).map(_.asDigit).toArray
+
+    Heightmap(width, height, data)
+end Heightmap
+
+
+def drawGrid(height: Int, width: Int, data: (Position, Height)*) =
+  val provided: Map[Position, Height] = data.toMap
+  var x = 0
+  var y = 0
+  val result = StringBuilder()
+  while (y < height) do
+    while (x < width) do
```

Comment:
Hum ... Why not
```suggestion
  for y <- 0 until height do
    for x <- 0 until width do
```
and get rid of the increments and assignments?

Reply:
I can do that, I do not have strong opinions on this. The function was used only to debug. Maybe I could even remove it